### PR TITLE
chore(1639) - Create github pipeline for SPA deployments with SPAship

### LINKED ISSUE #1639: Create github pipeline for SPA deployments with SPAship
## Is your chore related to a problem? Please describe

- No direct deployments

## Describe alternatives you've considered

- N/A

### PR DESCRIPTION
**Pull request by hybridx**
# Closes/Fixes/Resolves

- Fixes #1639 

# Explain the feature/fix

- - Added manual workflow

## Does this PR introduce a breaking change

- No

## Screenshots

N/A

### Ready-for-merge Checklist

- [ ] Expected files: all files in this pull request are related to one feature request or issue (no stragglers)?
- [ ] Did you update or add any necessary documentation (README.md, WHY.md, etc.)?
- [ ] Was this feature demoed and the design review approved?

### CODE REVIEW

**Comment by hybridx** [comment]
@riginoommen @akhilmhdh @deshmukhmayur Ignore the commits and squash merge the PR

**Review by akhilmhdh — APPROVED** [review]

**Review by gisumit — APPROVED** [review]

File: packages/home-spa/.env.example [hunk 1]
```diff
@@ -6,3 +6,4 @@ OP_SUBSCRIPTIONS_URL=
 
 ## SENTRY DSN
 # SENTRY_DSN=
+# Checking changed files
```
Comment (riginoommen) [inline packages/home-spa/.env.example:9]: Can you please remoce the duplicate env variable and comment

File: packages/home-spa/.env.example [hunk 2]
```diff
@@ -6,3 +6,4 @@ OP_SUBSCRIPTIONS_URL=
 
 ## SENTRY DSN
```
Comment (riginoommen) [inline packages/home-spa/.env.example:7]: Can you please remoce the duplicate env variable and comment

**Review by riginoommen — CHANGES_REQUESTED** [review]

File: .github/workflows/auto-deployments.yml [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        # Steps in the following script
+        # #1 cd into dir
+        # #2 install the required node_modules
+        # #3 get env file from gitlab (note: without the full url curl does not resolve the endpoint)
+        # #4 build with the .env file
+        # #5 deploy using spaship cli
+        run: |
+          cd ${{ matrix.container }}
+          npm install
+          curl -k --header "PRIVATE-TOKEN:${{ secrets.GITLAB_TOKEN }}" "https://gitlab.cee.redhat.com/api/v4/projects/9092/repository/files/spa-configs%2Fdev%2F${{ steps.get-spa-name.outputs.spa-name }}%2F.env/raw?ref=master" -o .env
```
Comment (riginoommen) [inline .github/workflows/auto-deployments.yml:71]: Can you use this in the environments because Its internal URL can you use in the environment variables

File: .github/workflows/auto-deployments.yml [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+          --env=$SPASHIP_DEPLOY_URL \
+          --ref=\$VERSION
+      - name: Build and deploy SPA on QA
+        if: contains(matrix.container, '-spa')
+        env:
+          SPASHIP_DEPLOY_URL: "${{ secrets.SPASHIP_DEPLOY_URL }}qa"
+        run: |
+          cd ${{ matrix.container }}
+          npm install
+          curl -k --header "PRIVATE-TOKEN:${{ secrets.GITLAB_TOKEN }}" ""https://gitlab.cee.redhat.com/api/v4/projects/9092/repository/files/spa-configs%2Fqa%2F${{ steps.get-spa-name.outputs.spa-name }}%2F.env/raw?ref=master" -o .env
```
Comment (riginoommen) [inline .github/workflows/auto-deployments.yml:84]: Can you use this in the environments because Its internal URL can you use in the environment variables

File: .github/workflows/manual-deployment.yml [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+        cache: 'npm'
+        cache-dependency-path: 'packages/**/package-lock.json'
+    - name: Build SPA with env
+      env:
+        CI: false
+        API_KEY: ${{ secrets.SPASHIP_API_KEY }}
+      run: |
+        cd packages/$SPA_NAME
+        npm install
+        curl -k --header "PRIVATE-TOKEN:${{ secrets.GITLAB_TOKEN }}" "https://gitlab.cee.redhat.com/api/v4/projects/9092/repository/files/spa-configs%2F${{ inputs.env }}%2F$SPA_NAME%2F.env/raw?ref=master" -o .env
```
Comment (riginoommen) [inline .github/workflows/manual-deployment.yml:46]: Can you use this in the environments because Its internal URL can you use in the environment variables

**Review by riginoommen — CHANGES_REQUESTED** [review]

File: packages/home-spa/.env.example (hunk 1 above)
Comment (hybridx) [reply packages/home-spa/.env.example:9]: @riginoommen both are required in home-spa

File: .github/workflows/auto-deployments.yml (hunk 3 above)
Comment (hybridx) [reply .github/workflows/auto-deployments.yml:71]: I did try using github secrets which didn't work, let me try again and check if it's possible

File: .github/workflows/auto-deployments.yml (hunk 4 above)
Comment (hybridx) [reply .github/workflows/auto-deployments.yml:84]: I did try using github secrets which didn't work, let me try again and check if it's possible

File: .github/workflows/manual-deployment.yml (hunk 5 above)
Comment (hybridx) [reply .github/workflows/manual-deployment.yml:46]: I did try using github secrets which didn't work, let me try again and check if it's possible

File: packages/home-spa/.env.example (hunk 2 above)
Comment (hybridx) [reply packages/home-spa/.env.example:7]: both are required

File: .github/workflows/auto-deployments.yml (hunk 4 above)
Comment (hybridx) [reply .github/workflows/auto-deployments.yml:84]: for example if you see this commit https://github.com/1-Platform/one-platform/pull/1642/commits/737a790035eee8a067ac9c1d5f05f29e1057c892 it was saved in github secrets and didn't work

File: packages/home-spa/.env.example (hunk 1 above)
Comment (hybridx) [reply packages/home-spa/.env.example:9]: One is the actual variable and other is the comment

File: .github/workflows/auto-deployments.yml [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+      #     SPASHIP_DEPLOY_URL: "${{ secrets.SPASHIP_DEPLOY_URL }}qa"
+      #   run: |
+      #     cd ${{ matrix.container }}
+      #     npm install
+      #     curl -k --header "PRIVATE-TOKEN:${{ secrets.GITLAB_TOKEN }}" ""https://gitlab.cee.redhat.com/api/v4/projects/9092/repository/files/spa-configs%2Fqa%2F${{ steps.get-spa-name.outputs.spa-name }}%2F.env/raw?ref=master" -o .env
+      #     npm run build
+      #     export VERSION=\$(node -p "require('./package.json').version")
+      #     npx @spaship/cli@latest deploy --apikey=${{ secrets.SPASHIP_API_KEY }} \
+      #     --env=$SPASHIP_DEPLOY_URL \
+      #     --ref=\$VERSION
```
Comment (riginoommen) [inline .github/workflows/auto-deployments.yml:97]:
```suggestion
      # - name: Build and deploy SPA on QA
      #   if: contains(matrix.container, '-spa')
      #   env:
      #     SPASHIP_DEPLOY_URL: "${{ secrets.SPASHIP_DEPLOY_URL }}qa"
      #   run: |
      #     cd ${{ matrix.container }}
      #     npm install
      #     curl -k --header "PRIVATE-TOKEN:${{ secrets.GITLAB_TOKEN }}" ""https://gitlab.cee.redhat.com/api/v4/projects/9092/repository/files/spa-configs%2Fqa%2F${{ steps.get-spa-name.outputs.spa-name }}%2F.env/raw?ref=master" -o .env
      #     npm run build
      #     export VERSION=\$(node -p "require('./package.json').version")
      #     npx @spaship/cli@latest deploy --apikey=${{ secrets.SPASHIP_API_KEY }} \
      #     --env=$SPASHIP_DEPLOY_URL \
      #     --ref=\$VERSION
```

remove the commented code if you are making it back active use ```GITLAB_URL``` as the env variable

**Review by riginoommen — CHANGES_REQUESTED** [review]

File: .github/workflows/auto-deployments.yml (hunk 6 above)
Comment (hybridx) [reply .github/workflows/auto-deployments.yml:97]: @riginoommen this secret isn't working

Comment (hybridx) [reply .github/workflows/auto-deployments.yml]: @riginoommen change added

File: .github/workflows/auto-deployments.yml (hunk 3 above)
Comment (hybridx) [reply .github/workflows/auto-deployments.yml]: Resolving as @riginoommen is on a PTO

File: .github/workflows/auto-deployments.yml (hunk 4 above)
Comment (hybridx) [reply .github/workflows/auto-deployments.yml]: Resolving as @riginoommen is on a PTO

**Review by soumyadip007 — COMMENTED** [review]
LGTM

**Review by gisumit — APPROVED** [review]

**Review by arkaprovob — COMMENTED** [review]
Looks good to me.

**Review by soumyadip007 — APPROVED** [review]

File: packages/home-spa/.env.example (hunk 1 above)
Comment (deshmukhmayur) [inline packages/home-spa/.env.example:9]: Can you remove this line?
Looks like something that was used to test the action job. Not something that should be merged.

**Review by akhilmhdh — APPROVED** [review]

File: .github/workflows/manual-deployment.yml [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+        type: choice
+
+env:
+  NODE_VERSION: 16
+  SPA_NAME: "${{ inputs.spaName }}-spa"
+  SPASHIP_DEPLOY_URL: "${{ secrets.SPASHIP_DEPLOY_URL }}${{ inputs.env }}"
+
+jobs:
+  build:
+   runs-on: self-hosted
```
Comment (deshmukhmayur) [inline .github/workflows/manual-deployment.yml:29]: Can you split this job into 2, in order to optimize it.
Since we don't need the self-hosted runner for building the app. Just for deploying.
So run the build on github shared runner, and the deploy on self-hosted.

**Review by deshmukhmayur — CHANGES_REQUESTED** [review]

File: .github/workflows/manual-deployment.yml (hunk 7 above)
Comment (hybridx) [reply .github/workflows/manual-deployment.yml:29]: @deshmukhmayur the SPA needs to be built on the self hosted container as it needs the env files which are also internal

File: packages/home-spa/.env.example (hunk 1 above)
Comment (hybridx) [reply packages/home-spa/.env.example:9]: Needed for testing after merge is done

File: .github/workflows/manual-deployment.yml (hunk 7 above)
Comment (deshmukhmayur) [reply .github/workflows/manual-deployment.yml:29]: Oh right. Nevermind that then.
